### ISSUE TITLE
Mapper shootout: sklearn baseline vs TabFM vs needle3 on the enum-mapping job

### ISSUE THREAD

**Issue by Ravicha2** [Ravicha2/transaction-extractor#12]
## What to build

The #11 hybrid isolated the only model job left: (vendor, printed bank label) → frozen enum. needle3 zero-shot scores **40%** there. This issue runs the same eval on progressively stronger mappers so the REPORT verdict — 'do you need a foundation model for this at all?' — falls out of numbers:

1. **sklearn arm (build first)** — TF-IDF char-n-grams over `vendor + printed` → LogisticRegression, trained on the labeled history. Training pairs are derived by joining the deterministic extraction (#11) to `ground_truth.csv` on (date, amount) — no relabeling. **Scoring protocol: leave-one-out** — each labeled transaction is predicted by a mapper fit on the other 9 — because fitting on the eval rows and resubmitting would inflate the score. Non-matching records (unscored extras) are predicted with the full fit.
2. **TabFM arm (second)** — google-research/tabfm, scikit-learn-compatible tabular ICL; worth trying before Jev since it is free (weights are non-commercial-licensed, fine for personal use) and local. Honest caveat: 10 context rows is deep below its design point; it earns its keep only if it generalizes to unseen vendors.
3. **Jev (stretch, stays #8)** — likely the precision ceiling (choice-over-enum is its primitive) but not free; run last if at all.

All arms feed the existing #7 harness unchanged: same matching, same bars (category ≥85%), same cohort breakdown. The question each number answers: is this job memorization (lookup/LR suffices), in-context tabular (TabFM), or semantic (needs a real LM)?

## Acceptance criteria

- [ ] `scripts/run_sklearn_arm.py`: one command writes records to `data/out/sklearn/` in the shared schema, LOO-flagged
- [ ] #7 harness scores the arm without changes beyond adding the directory
- [ ] Category accuracy vs the 85% bar, reported next to needle3's 40%
- [ ] TabFM arm attempted after the sklearn number lands (torch/JAX dep noted)

## Blocked by

- Nothing (builds on #11's extraction and #7's harness)

**Comment by Ravicha2** [comment]
## sklearn arm built and scored: LOO 0%, resubmission 100% — and that gap IS the finding

`scripts/run_sklearn_arm.py`: reuses the #11 hybrid's deterministic extraction verbatim, derives training pairs by joining extracted records to `ground_truth.csv` on (date, amount) — no relabeling — then TF-IDF char-wb 2–4 grams + LogisticRegression over `vendor + printed`. Scored by the #7 harness unchanged (`--arm sklearn`).

**Numbers (10 labeled keys / 16 labeled rows):**

| protocol | category accuracy | note |
|---|---|---|
| group LOO (honest) | **0%** (0/10) vs ≥85% bar | FAIL |
| resubmission (full fit, train-set) | **100%** (16/16) | upper bound |

**Why LOO is 0% — structure, not model choice.** Group holdout removes all rows of one label key; three classes — `transport`, `tuition fee`, `deposit` — exist as a *single key each* (the same transaction re-shown across screenshots), so for 3 of 10 scored labels the class is **absent from training by construction**: guaranteed wrong before the model runs. The remaining rows fail because char-n-grams over 16 short strings share almost no features — the classifier collapses to class priors. Meanwhile resubmission is perfect: **memorization of recurring vendors works flawlessly; there is not enough labeled history yet for any trained mapper to prove it honestly.**

**Verdict ladder this sets up for REPORT.md:**
- Extraction (date/amount/vendor/printed): regex+deterministic, 100% recovery — solved, no model.
- Mapping: needle3 zero-shot **40%** (data-free, can't beat priors), sklearn trained **0% honest / 100% ceiling** (data-starved), and both TabFM and Jev would hit the *same structural wall today*: TabFM's context would also lose the held-out class under fair LOO with 10 keys.

**Recommendation:** pause the TabFM arm until the label set covers the full ~30-screenshot corpus (~100+ rows, multiple keys per class). At 10 labels it is a foregone conclusion; at 100+ it becomes the real test of in-context tabular generalization vs a plain lookup — which is the comparison worth paying torch's install for. Jev (#8) stays last regardless (not free).

**Comment by Ravicha2** [comment]
## All current numbers are provisional — rerun required on the full corpus

Everything scored so far (needle3 zero-shot 40%, sklearn LOO 0% / resubmission 100%, the calibration overlap) is on a **4-screenshot corpus with 10 labeled keys / 16 labeled rows**. That is far below the scope's ~30-screenshot corpus, and the #12 diagnosis shows the numbers are data-size-dominated: classes held out wholesale by group-LOO, n-grams with nothing to share.

**When the remaining ~26 screenshots are OCR'd (#2's batch) and labeled (#3's label set extended), rerun before drawing any verdict:**

```
uv run python scripts/run_ocr.py
uv run python scripts/run_needle3.py          # zero-shot arm
uv run python scripts/run_hybrid_arm.py       # needle3 mapper arm
uv run python scripts/run_sklearn_arm.py      # sklearn mapper arm
uv run python scripts/run_eval.py --arm regex
uv run python scripts/run_eval.py --arm needle3
uv run python scripts/run_eval.py --arm hybrid
uv run python scripts/run_eval.py --arm sklearn
```

(all arms are resume-safe; pass `--force` to redo). The harness and arm scripts are frozen so reruns are directly comparable. Expect the sklearn/TabFM LOO numbers to move the most — at ~100+ rows every class has multiple label keys and the structural guaranteed-wrong cohort disappears. TabFM arm proceeds then too. Same requirement propagates to #9 (calibration table) and #10 (report must either use full-corpus numbers or flag these as preliminary n=10).

**Comment by Ravicha2** [comment]
## Full-corpus rerun on the extended corpus + updated ground truth: sklearn 79.4% LOO vs needle3-mapper 37.5% — memorization wins so far, but nobody clears 85%

Corpus now: **10 screenshots, 40 label rows / 39 keys, 7 classes**. The enum was amended to add `withdraw` (verbatim from the updated `ground_truth.csv`, same rule as the first amendment; the hybrid's enum-definition line updated to match). All arms re-run `--force`, so every number below is same-enum, same-GT, same harness (#7 unchanged).

| arm | record recovery (≥90%) | record precision | category accuracy (≥85%) |
|---|---|---|---|
| regex | 85.0% (34/40) FAIL | 53.1% | n/a |
| needle3 zero-shot | 0.0% (0/40) FAIL | 0.0% | n/a — emitted 0 records on 6/10 screenshots |
| hybrid (regex + needle3 mapper) | 85.0% FAIL | 52.3% | **37.5%** (12/32) FAIL |
| sklearn (TF-IDF+LR, group LOO) | 85.0% FAIL | 52.3% | **79.4%** (27/34) FAIL |
| sklearn, resubmission (full fit) | — | — | 79.1% (34/43) |

**The mapper-shootout answer at 39 keys:** the trained mapper beat the zero-shot mapper by 42 points — needle3-as-mapper got *worse* with more data (40% → 37.5%, same "Misc and souvenir" repetition attractor: Income→Misc, Education→Misc still fail; only the lexical bridges work), while sklearn went 0% → 79.4%. All 7 sklearn confusions are minority classes collapsing to the grocery prior: `deposit`, `tuition fee`, `withdraw` (1 key each — structurally unwinnable under group LOO), plus health×2, transport×1, Misc×1. And the resubmission ceiling now *equals* LOO (79.1 vs 79.4): at this size TF-IDF+LR can't even memorize its own training rows, so honest score ≈ ceiling. Verdict so far: the mapping job is prior + memorization; no evidence yet that semantics are needed — but no arm clears the 85% bar.

**The recovery cap is structural, and it's a GT-date-semantics issue — needs a ruling.** `ocr_miss=0`: every one of the 6 missing amounts IS in some OCR text. The updated ground truth dates these rows to their **true transaction dates**, while the bank app groups them under the **posting date** the screenshot displays:

| GT (true date) | transaction | displayed under |
|---|---|---|
| 16/9 −11.03 | Vodafone (Utilities) | Fri 28 Aug |
| 16/9 −46.63 | Humanitix (Gifts & donations) | Tue 1 Sep |
| 16/9 −42.50 | UNSW Fitness & Aquatic | Thu 10 Sep |
| 10/9 −3.30 | Transport NSW | Wed 16 Sep |
| 1/9 −7.99 | IGA (Kensington) | Wed 16 Sep |
| 28/8 −24.99 | Chemist Warehouse | Wed 16 Sep |

Every extraction arm reads the screenshot faithfully (that's the spec: "resolved from sticky headers + capture timestamp"), so these 6 are unwinnable by construction and cap recovery at 85% for **every** arm, needle3 included. Until resolved, the 90% recovery bar is un-passable by design. Options: (a) GT reverts to posting dates as displayed, or (b) the harness accepts the displayed posting date as a match for a true-date label.

**needle3 zero-shot on the new screenshots: 0 records on 6/6.** All six new screenshots use the two-column list layout (amounts vertically detached from vendor lines); the frozen input shape (y-sorted OCR lines) interleaves the columns into noise and the grammar decode collapses. Consistent with the #5/#6 attachment finding, now at corpus scale.

**TabFM gate: still holds.** 39 keys is well under the ~100+ gate from the previous comment, and 3 classes still have exactly 1 key — guaranteed wrong under fair group LOO for any in-context mapper too. Planned shape when it runs, for the record: one row per transaction, feature columns = vendor + printed bank label (+ derived per-transaction date features such as day-of-week; raw dates would memorize the calendar and leak the label join), target = enum category; TabFM retrieves k similar training rows as the in-context table per query, with the same group-LOO exclusion.

**Comment by Ravicha2** [comment]
## TabFM declared operationally infeasible; Jev arm (#8) completes the shootout: Jev 94.1% PASS — the only arm over the 85% bar

### TabFM arm: attempted, declared infeasible on personal hardware

`scripts/run_tabfm_arm.py` is built and committed for the record, but never scored by protocol. Measured on this machine (torch CPU float32, `google/tabfm-1.0.0-pytorch` weights): predict = a forward pass of the 24-block ICL transformer over the whole context table **per ensemble member** — ~25-40 s per call, mostly fixed cost (1 vs 8 query rows barely differs; fit is just encoding at 0.7 s). Group LOO needs one call per label group (33 groups + 1 extras batch = 34 calls) → **20-30 min per rerun**, for a job the sklearn arm does in ~50 ms and needle3 decodes in seconds. KV-cache-style amortization does not apply: within a call the context is already shared; across calls every LOO group has a *different* 42-row context and attention is bidirectional, so cached context representations would be wrong — zero hit rate by construction. MPS benchmarked slower than CPU at this sequence length. An `n_estimators=1` fallback (~3 s/call) was scoped but superseded by Jev before scoring. **The cost asymmetry is itself the finding: in-context tabular has no training to amortize — the wrong cost shape for a 65-record personal corpus.**

### Jev arm (official TypeSafe API, per owner direction — not the openjev repro)

`scripts/run_jev_arm.py`: one Choice question per transaction at `api.typesafe.ai/v1/systemone` (`jev-latest`, jev-1.13.0). State = vendor + printed label (identical text the #11 hybrid mapper sees); criteria = the frozen enum strings verbatim with the same meaning definitions. Zero-shot, **data-free** — labels never enter, so no LOO protocol; every record predicted, all outputs flagged `ood=true` per #8. 65 calls, ~1 minute, fractions of a cent.

### The completed shootout ladder (full corpus, 40 label rows, 7 classes)

| mapper (same extraction, same labels) | category accuracy (≥85%) | notes |
|---|---|---|
| needle3 121M, per-record mapping | 37.5% (12/32) | repetition attractor; Income/Education semantically-mediated maps fail |
| sklearn TF-IDF+LR, group LOO | 79.4% (27/34) | resubmission ceiling ≈ LOO (79.1%) — can't even memorize at 43 rows |
| **Jev (jev-latest), zero-shot** | **94.1% (32/34) PASS** | OOD-flagged; only arm over the bar |

Jev wins exactly where the others structurally fail — the minority classes: `Income → deposit`, `Education → tuition fee`, and the `Withdrawal CBA ATM` row → `withdraw` at p=1.0 (the deterministic-lexical row every trained mapper misses under group-LOO). Its only 2 confusions are the 26/8 Regent Hotel rows (printed "Entertainment", labeled `grocery&eat out`) → `Misc and souvenir`, arguably the label being the odd one.

**Verdict for REPORT.md:** the enum-mapping job is *semantic*, not memorization — a trained n-gram mapper is capped by its representation (ceiling = LOO ≈ 79%, minority classes guaranteed wrong at 1 key each), while a zero-shot typed-choice model clears the bar data-free at ~100 ms/record. Foundation-model cost is justified for the mapping step only; extraction stays deterministic regex (85% recovery, capped by the GT posting-date-vs-true-date issue that still needs a ruling — unchanged from the previous comment, `ocr_miss=0`).
